File: fpm_core/test_fpm.cpp
// test_fpm.cpp — Stage 4: chạy CFastMatch trên ảnh thật (Test Images/Src1+Dst1),
// so sánh nhánh SIMD on vs off. PASS nếu cả 2 đều match và top result trùng nhau.
#include "fpm_core.h"
#include <opencv2/imgcodecs.hpp>
#include <cstdio>

using namespace fpm;
using cv::Mat;

static void run (bool bSIMD, const Mat& src, const Mat& templ, std::vector<s_SingleTargetMatch>& out)
{
	MatchParams params;
	params.bUseSIMD = bSIMD;
	CFastMatch matcher;
	matcher.LearnPattern (templ, params);
	out = matcher.Match (src, params);
}

int main (int argc, char** argv)
{
	const char* srcPath = argc > 1 ? argv[1] : "Test Images/Src1.bmp";
	const char* dstPath = argc > 2 ? argv[2] : "Test Images/Dst1.bmp";

	Mat src = cv::imread (srcPath, cv::IMREAD_GRAYSCALE);
	Mat templ = cv::imread (dstPath, cv::IMREAD_GRAYSCALE);
	if (src.empty () || templ.empty ())
	{
		printf ("Cannot load images: %s / %s\nSTAGE4 FAIL\n", srcPath, dstPath);
		return 2;
	}
	printf ("src=%dx%d templ=%dx%d\n", src.cols, src.rows, templ.cols, templ.rows);

	std::vector<s_SingleTargetMatch> r0, r1;
	run (false, src, templ, r0);
	run (true, src, templ, r1);

	auto dump = [] (const char* tag, const std::vector<s_SingleTargetMatch>& r) {
		printf ("[%s] matches=%zu", tag, r.size ());
		if (!r.empty ())
			printf ("  top: center=(%.2f,%.2f) angle=%.2f score=%.4f", r[0].ptCenter.x, r[0].ptCenter.y, r[0].dMatchedAngle, r[0].dMatchScore);
		printf ("\n");
	};
	dump ("SIMD=0", r0);
	dump ("SIMD=1", r1);

	int rc = 0;
	if (r0.empty () || r1.empty ()) { printf ("[FAIL] một nhánh không match\n"); rc = 1; }
	else
	{
		double dc = cv::norm (r0[0].ptCenter - r1[0].ptCenter);
		double da = std::abs (r0[0].dMatchedAngle - r1[0].dMatchedAngle);
		printf ("diff top: dCenter=%.3f px, dAngle=%.3f deg\n", dc, da);
		if (dc > 2.0 || da > 1.0) { printf ("[FAIL] 2 nhánh SIMD lệch nhau\n"); rc = 1; }
	}
	printf (rc == 0 ? "STAGE4 PASS\n" : "STAGE4 FAIL\n");
	return rc;
}

File: fpm_core/test_fpm.py
"""Test for the fpm module (.pyd) — load images with PIL, call match, print results.

Run: python test_fpm.py
DLL search paths (fpm.pyd + OpenCV) are configured by fpm_env before importing fpm.
"""
import os
import sys
import numpy as np
from PIL import Image

from fpm_env import setup_fpm_dll_dirs, HERE

setup_fpm_dll_dirs()
import fpm  # noqa: E402

# Sample images: local copy in fpm_core, else the parent project's "Test Images".
_IMG_DIRS = [os.path.join(HERE, "Test Images"),
             os.path.join(os.path.dirname(HERE), "Test Images")]
IMG_DIR = next((d for d in _IMG_DIRS if os.path.isdir(d)), HERE)

def load_gray(path):
    return np.asarray(Image.open(path).convert("L"))

def main():
    src = load_gray(os.path.join(IMG_DIR, "Src1.bmp"))
    tmpl = load_gray(os.path.join(IMG_DIR, "Dst1.bmp"))
    print(f"src={src.shape} templ={tmpl.shape} dtype={src.dtype}")

    params = fpm.MatchParams()
    params.score = 0.5
    params.max_pos = 10
    print("params:", params)

    matcher = fpm.FastMatch()
    matcher.learn(tmpl, params)
    print("learned:", matcher.is_learned())
    res = matcher.match(src, params)

    print(f"\n{len(res)} match:")
    for i, m in enumerate(res):
        cx, cy = m["center"]
        print(f"  #{i}: score={m['score']:.4f} angle={m['angle']:.2f} "
              f"center=({cx:.1f},{cy:.1f})")

    ok = len(res) > 0 and res[0]["score"] > 0.9
    print("\nPYTEST", "PASS" if ok else "FAIL")
    sys.exit(0 if ok else 1)

if __name__ == "__main__":
    main()
